### ISSUE TITLE
TypeScript 5.0 decorators support

### ISSUE THREAD

**Issue by d0rich** [typeorm/typeorm#9862]
### Feature Description

In TypeScript 5.0 decorators will be changed, but it will become stable feature (https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-beta/). 

Are you going to support new decorators?

### The Solution

Use TypeORM in project with new TypeScript decorators. (I want create my own decorators in other layers of program)

### Considered Alternatives

Use legacy decorators in projects with TypeScript 5.0

### Relevant Database Driver(s)

- [ ] aurora-mysql
- [ ] aurora-postgres
- [ ] better-sqlite3
- [ ] cockroachdb
- [ ] cordova
- [ ] expo
- [ ] mongodb
- [ ] mysql
- [ ] nativescript
- [ ] oracle
- [ ] postgres
- [ ] react-native
- [ ] sap
- [ ] spanner
- [ ] sqlite
- [ ] sqlite-abstract
- [ ] sqljs
- [ ] sqlserver

### Are you willing to resolve this issue by submitting a Pull Request?

No, I don’t have the time and I’m okay to wait for the community / maintainers to resolve this issue.

**Comment by pleerock** [comment]
Yes, we will support new standard. Contributions are welcomed for this feature. However we need to avoid breaking changes.

**Comment by lee-at-built** [comment]
Hey @pleerock how will you support the new standard if TS does not emit static type information to the decorators as before (via the now deprecated `emitDecoratorMetadata` option)? It seems essential for TypeORM to know the types of various fields at runtime (to know a column is a string vs number etc).

**Comment by pleerock** [comment]
users will be able to specify the type explicitly in the column options.

**Comment by davecarlson** [comment]
Game changing this! Going to simplifty bulid workflows massively

**Comment by akinorimizushima** [comment]
looks like they will include the decorator metadata to v5.2
https://github.com/microsoft/TypeScript/issues/54298

**Comment by whitehat101** [comment]
Does that mean that TypeORM should work with TypeScript 5.2? Or is there work to be done to support 5.2 decorators?

**Comment by markedwards** [comment]
Is this is official issue to track this?  Is there any progress or momentum on it?

**Comment by trusktr** [comment]
@pleerock can you please chime in on this problem with the new standard decorators, and the proposed solution, and whether the current design makes it harder to migrate to new decorators (not only for implementation, but for users too), and whether the solution idea would make things easier?

[discrepancy: we're allowed to get/set private fields, but not public fields, and public auto accessors are not fields.](https://x.com/trusktr/status/1846989651543802133)

**Comment by ImBoop** [comment]
This should probably be prioritized as this would gate against using TypeORM in any project that relies on the new decorators or any environment where you can't otherwise enable legacy decorators.

**Comment by ImBoop** [comment]
This issue is pretty critical for the longevity of TypeORM, I'd suggest it get reopened

**Comment by hershmire** [comment]
Is there any planned work for this? It looks like Typescript v6 is deprecating experimentalDecorators which will make it a blocker to eventually transition to v7.

**Comment by hershmire** [comment]
Typescript v7 is now out. Is there any update on where this stands?

**Comment by alumni** [comment]
> Typescript v6 is deprecating experimentalDecorators

This is not true, the [experimental decorators](https://www.typescriptlang.org/tsconfig/#experimentalDecorators) are not deprecated and are not going anywhere anytime soon. Deprecated features would have a deprecated label in the TS documentation.

We will consider it the future, but to be able to implement them we will need to do some refactoring and split TypeORM into packages. We plan to do this refactor for related reasons (to make TypeORM more extensible) in the near future anyway, but the first priority would be driver and platform support.

If you don't want to use decorators, you can define entities using [entity schemas](https://typeorm.io/docs/entity/separating-entity-definition/#defining-schemas).

**Comment by jacobg** [comment]
> > Typescript v6 is deprecating experimentalDecorators
> 
> This is not true, the [experimental decorators](https://www.typescriptlang.org/tsconfig/#experimentalDecorators) are not deprecated and are not going anywhere anytime soon. Deprecated features would have a deprecated label in the TS documentation.
> 
> We will consider it the future, but to be able to implement them we will need to do some refactoring and split TypeORM into packages. We plan to do this refactor for related reasons (to make TypeORM more extensible) in the near future anyway, but the first priority would be driver and platform support.
> 
> If you don't want to use decorators, you can define entities using [entity schemas](https://typeorm.io/docs/entity/separating-entity-definition/#defining-schemas).

We used to use the experimental decorators, but removed it a few years ago because it was causing various bundling issues, although I forget the details. I think it's sort of a neglected feature in build tools, since people view it as something that is a dead end due to being rejected by ecmacscript standards body.

We switched to using entity schemas, but that is also not ideal, because it doesn't provide a way to define the various hooks that the decorators allow.
